### PR TITLE
Add CLI to plot memory monitors

### PR DESCRIPTION
**Pull request by knelli2**
## Proposed changes

Examples of output:

[no_color.pdf](https://github.com/sxs-collaboration/spectre/files/14397432/no_color.pdf)

[with_color.pdf](https://github.com/sxs-collaboration/spectre/files/14397430/with_color.pdf)

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Visualization/Python/CMakeLists.txt
Comment (knelli2) [inline tests/Unit/Visualization/Python/CMakeLists.txt]: For some reason, the command itself takes a very long time to run. I wasn't able to determine why, even with `spectre --profile`. Any suggestions of where I'm going wrong are welcome.

**Comment by nilsdeppe** [comment]
I'll let @nilsvu review the actual code, but I have a feature request: would it be easy to add a flag like `--top-n 5` to show only the 5 most significant users? On the plots you showed it was tricky to figure out from the legend which were the 3 big ones shown.

**Comment by nilsvu** [comment]
Good idea, and/or how about a stacked area plot like this: https://python-graph-gallery.com/stacked-area-plot/
Then only show the legend entry for the big ones (above 5% or something like that)

**Comment by nilsdeppe** [comment]
I'd be fine with adding that as an option. I do find it easier in the current style to read off how many GB something is using. I think the stacked plot makes relative comparisons nice, though. So maybe a CLI flag to enable the stacking? E.g. `--stacked-plot`?

**Comment by knelli2** [comment]
I'm personally not a big fan of the stacked plot. It's a bit ambiguous to me whether to interpret a stack as relative to zero, or relative to the stack below it. I do like the `--top-n` option though. I'll add that.

**Comment by nilsvu** [comment]
True. Maybe just combine the entries with only a couple percent into a single entry, then you don't need an extra CLI option and you can't see those small entries in the plot anyway. Also sort the entries.

**Comment by knelli2** [comment]
Examples of output now with fixup

no_color

with_color

**Comment by nilsdeppe** [comment]
LGTM, once @nilsvu is happy with the python code, please go ahead and merge. No need for my approval.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py:39]: I'd remove this. Either make it always GB or switch to MB if the total memory is below 1 GB. This is just another option the person using this has to learn and set.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: Also remove this. The stylesheet sets the color cycle (see PlotPowerMonitors.py how to evaluate the color cycle if needed).

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: There's no way to distinguish these lines (no legend, no colors or other property cycles). Just delete the nocolor option and have the stylesheet define the style.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: use `np.argsort` to get the permutation. you can also do this earlier and then plot in the right order so you don't have to reorder the legend.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: Maybe `total_over_nodes`?

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: You read every H5 file multiple times, once for each parallel component. Not sure how much data we'll deal with here, but to fix this and read each h5file only once you can do this: `pd.concat(pd.DataFrame({subfile: total_over_nodes(to_dataframe(subfile)) for subfile in memory_filenames}) for h5file in h5files)` (pseudocode, maybe not do this in 1 line :D). This also gets you a single DataFrame of all parallel components over time, so you don't have to deal with dicts anymore.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (knelli2) [reply src/Visualization/Python/PlotMemoryMonitors.py:39]: I'd prefer to keep this one. I did make the default GB though since that's more likely

Comment (nilsvu) [reply src/Visualization/Python/PlotMemoryMonitors.py:39]: Ok

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: You do this only once, so you don't need the function I think? Also update the comment about "dicts" above.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: This is just `total = df.sum(index=1)`

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]:
```py
rest_threshold = 0.01 * total.mean()
components_to_plot = df.columns[df.max() > rest_threshold]
rest_columns = [col for col in df.columns if col not in cols]
the_rest = df[rest_columns].sum(axis=1)
```

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: (optional)
```py
maxes = totals_df[components_to_plot].max()
components_to_plot = components_to_plot[np.argsort(maxes)[::-1]]
```

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py:199]: Is the line width needed because the time sampling is so dense? Generally better to leave style up to the stylesheet.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: `return to_dataframe(subfile).set_index("Time")` makes everything below easier

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: Want to sum over all columns? Then just do `df.sum(axis=1)`. Otherwise `df[cols_to_sum].sum(axis=1)`. Suggestion:

```py
def total_over_nodes(df):
    cols_to_sum = [col == "Size (MB)" or "Size on node" in col for col in df.columns]
    return df[cols_to_sum].sum(axis=1)
```

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (nilsvu) [inline src/Visualization/Python/PlotMemoryMonitors.py]: can delete this

**Review by nilsvu — CHANGES_REQUESTED** [review]
Some simplifications with pandas. You can squash everything.

File: src/Visualization/Python/PlotMemoryMonitors.py
Comment (knelli2) [reply src/Visualization/Python/PlotMemoryMonitors.py:199]: I did this so the total line would stand out in comparison to the other lines

Comment (nilsvu) [reply src/Visualization/Python/PlotMemoryMonitors.py:199]: Ah ok, maybe the color is enough? Up to you.

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
